packages/chevrotain/src/parse/parser/smart_parser/smart_recognizer_engine.ts

```diff
@@ -0,0 +1,161 @@
+import {
+    ConsumeMethodOpts,
+    DSLMethodOpts,
+    GrammarAction,
+    IAnyOrAlt,
+    ManySepMethodOpts,
+    OrMethodOpts,
+    SubruleMethodOpts,
+    TokenType
+} from "../../../../api"
+import {
+    Repetition,
+    Terminal,
+    NonTerminal,
+    Alternation,
+    Flat,
+    Option
+} from "../../grammar/gast/gast_public"
+import { RecognizerEngine } from "../traits/recognizer_engine"
+import { SmartMixedInParser } from "./traits"
+import {
+    PROP_ANALYSER_CONTEXT,
+    IRuleDefinitionInfo,
+    ParserRuleAnalyserContext
+} from "./rule-analyser"
+import { MixedInParser } from "../traits/parser_traits"
+
+function TODO() {
+    throw new Error("TODO: not implemented yet")
+}
+
+/**
+ * This trait is responsible for the runtime parsing engine
+ * Used by the official API (recognizer_api.ts)
+ */
+export class SmartRecognizerEngine extends RecognizerEngine {
+    // Implementation of parsing DSL
+    optionInternal<OUT>(
+        this: SmartMixedInParser,
+        ...args: Parameters<RecognizerEngine["optionInternal"]>
+    ): any {
+        const ctx = this[PROP_ANALYSER_CONTEXT]
+
+        if (!ctx) {
+            return (super.optionInternal as any)(...args)
+        }
+
+        const [actionORMethodDef, occurrence] = args
+
+        if (typeof actionORMethodDef !== "function") {
+            return TODO()
+        }
+
+        const optionImpl = actionORMethodDef
+
+        ctx.addToRecorded(
+            new Option({
+                definition: ctx.recordDefinitions(optionImpl),
+                idx: occurrence
+            })
+        )
+    }
+
+    manyInternal<OUT>(
```

Comment:
Overriding the ***Internal methods assumes the parsing methods (MANY0-9) do not
perform anything else of consequences, this is probably true now, but would it always be true?
It could be o.k. to make this assumption for convenience instead of overriding dozens of methods, but we need to document this.